### DISCUSSION TITLE
Ability to validate structs against their jsonschema-tags

### DISCUSSION THREAD

**Discussion by benbender** [swaggest/rest#178] [Ideas]
I've tried to test usecases in isolation without the http-transport and request-decoding. This is easy but misses the schema validation. The validation seems to be bound to the the http-request if I did not miss something. This problem would also be the case if I would try to reuse a usecase f.e. in a CLI-context.

To me it seems like mapping and validation are two separate concerns. Ideally I would be able to `Interact` with a `usecase` and the validation would be done automagically as, at least to me, it's part of the business logic.

f.e.:
```go
type ValueInput struct {
	PhoneNr   string `json:"phone_nr" db:"phone_nr" required:"true" pattern:"^\\+[1-9]\\d{1,14}$"`
	FirstName string `json:"first_name" db:"first_name" minLength:"1" required:"true"`
	LastName  string `json:"last_name" db:"last_name" minLength:"1" required:"true"`
}

func TestCreate(t *testing.T) {
	ctx := context.Background()
	var out Account
	
	// manual "mapping"
	in := ValueInput{
		PhoneNr:   "4934226490",
		FirstName: "Foo",
		LastName:  "Bar",
	}

	// Create is a normal usecase.Interactor
	err := Create().Interact(ctx, in, &out)

	// Here I would expect an error about the wrong format of the PhoneNr missing the "+" at the beginning
	assert.Error(t, err)
}
```

Alternatively an API like this would also be fine so that I would be able to validate on a case by case basis:
```go
type ValueInput struct {
        ID int `json:"id" path:"accountID" required:"true"`
	PhoneNr   string `json:"phone_nr" db:"phone_nr" required:"true" pattern:"^\\+[1-9]\\d{1,14}$"`
	FirstName string `json:"first_name" db:"first_name" minLength:"1" required:"true"`
	LastName  string `json:"last_name" db:"last_name" minLength:"1" required:"true"`
}

// manual "mapping"
v := ValueInput{
        ID: 1,
	PhoneNr:   "4934226490",
	FirstName: "Foo",
	LastName:  "Bar",
}

validator := JSONSchemaValidator.New()
// Here I would expect an error about the wrong format of the PhoneNr missing the "+" at the beginning
err := validator.ValidateStruct(v)
```

**Comment by vearutop** [comment]
I like this idea, and also feels easy to implement.

One potential caveat here is lack of actual JSON in some cases, for example tag-based validation is applied also to `query` parameters that are artificially converted to JSON with a map (just for sake of validation). 

Perhaps such scenario might be relevant for other transports, or if we want to stick to `json` only, that would be easier.

**Reply by benbender** [reply]
Tbh, I don't fully grasp your concern here. From my viewpoint, the user would need to provide the `map` accordingly to the struct because mapping is his/her concern in these usecases (and the compiler would also spill if you would provide a query-string instead of a map f.e.).

The `query`-tag (and other mapping-related tags) would be ignored by the validator and only be used by the `request-decoder`/`mapping`-layer which resides inside the http/rest-context.

**Comment by ekristen** [comment]
I like this idea, especially for testable code.

I'd also like to see if it were possible to set the validation values via methods on the struct.

For example if I have an enum type, I would rather reference a string slice in my code as possible enums then have to update a tag that's static, or let's say I have a setting that allows customization of a max length of a field that an end user can set, being able to pull that in from a configuration an set maxlength to this dynamic setting on server startup when the openapi config is generated would be very nice.

Additionally, it would be nice if I could add `func (i *Input) Validate() error` to my input for further validation.

For example if I'm using a ULID or a UUID field, it would be nice to validate them as valid parseable values before getting into the interact portion of my code. 

@vearutop thoughts?
